lib/Sema/CSApply.cpp

```diff
@@ -2630,23 +2649,48 @@ namespace {
         // before taking a single element.
         baseTy = cs.getType(base);
         if (!toType->hasLValueType() && baseTy->hasLValueType())
-          base = coerceToType(base, baseTy->getRValueType(), cs.getConstraintLocator(base));
+          base = coerceToType(base, baseTy->getRValueType(),
+                              cs.getConstraintLocator(base));
 
         return cs.cacheType(new (cs.getASTContext())
                             TupleElementExpr(base, dotLoc,
                                              selected.choice.getTupleIndex(),
                                              nameLoc.getBaseNameLoc(), toType));
       }
 
-      case OverloadChoiceKind::BaseType: {
+      case OverloadChoiceKind::BaseType:
         return base;
-      }
 
       case OverloadChoiceKind::KeyPathApplication:
         llvm_unreachable("should only happen in a subscript");
+          
+      case OverloadChoiceKind::DynamicMemberLookup: {
+        // Application of a DynamicMemberLookup result turns a member access of
+        // x.foo into x[dynamicMember: "foo"].
+        auto subscriptDecl = cast<SubscriptDecl>(selected.choice.getDecl());
+
+        // Build and type check the string literal index value to the specific
+        // string type expected by the subscript.
+        Expr *nameExpr = new (cs.getASTContext())
+          StringLiteralExpr(selected.choice.getName().getBaseIdentifier().str(),
+                            nameLoc.getStartLoc(), /*implicit*/true);
+        auto nameLabel = cs.getASTContext().getIdentifier("dynamicMember");
+        
+        auto stringType = subscriptDecl->getIndices()->get(0)->getTypeLoc();
+        (void)cs.TC.typeCheckExpression(nameExpr, dc, stringType,
```

Comment:
What if this fails?

Reply:
This can't fail, because previous filtering has validated that the index type is ExpressibleByStringLiteral.  This is similar to the logic used forming string literals themselves.